src/OptimalControlUtilities/include/BipedalLocomotionControllers/OptimalControlUtilities/ControlProblemElements.h

```diff
@@ -0,0 +1,745 @@
+     */
+    void setOrientationPIDGains(const double& c0, const double& c1, const double& c2);
+
+    /**
+     * Set the linear PID gains
+     * @param kp proportional gain
+     * @param kd derivative gain
+     * @throw std::runtime_error if the Type is different from ONE_DIMENSION
+     */
+    void setOneDegreePIDGains(const double& kp, const double& kd);
+
+    /**
+     * Get (and compute) the element matrix
+     * @return the element matrix
+     */
+    virtual const iDynTree::MatrixDynSize& getA() final;
+
+    /**
+     * Get (and compute) the element vector
+     * @return the element vector
+     */
+    virtual const iDynTree::VectorDynSize& getB() final;
+
+    /**
+     * Set if the link is in contact with the environment
+     * @param isInContact true if the link associated to the frame is in contact with the
+     * environment
+     */
+    void isInContact(bool isInContact);
+};
+
+/**
+ * SystemDynamicsElement describes the System Dynamics that will be embedded as
+ * equality constraint or cost function.
+ */
+class SystemDynamicsElement : public CostFunctionOrEqualityConstraintElement
+{
+    /** Index range of the base acceleration */
+    iDynTree::IndexRange m_baseAccelerationIndex{iDynTree::IndexRange::InvalidRange()};
+
+    /** Index range of the joint acceleration */
+    iDynTree::IndexRange m_jointAccelerationIndex{iDynTree::IndexRange::InvalidRange()};
+
+    /** Index range of the joint torque */
+    iDynTree::IndexRange m_jointTorqueIndex{iDynTree::IndexRange::InvalidRange()};
+
+    iDynTree::MatrixDynSize m_massMatrix; /**< Floating-base mass matrix  */
+    iDynTree::FreeFloatingGeneralizedTorques m_generalizedBiasForces; /**< Coriolis and
+                                                                         Gravitational term  */
+    iDynTree::MatrixDynSize m_jacobianMatrix; /**< Jacobian Matrix  */
+
+    std::vector<Frame> m_framesInContact; /**< Vectors containing the frames in contact with the
+                                             environment */
+
+    iDynTree::MatrixDynSize m_reflectedInertia; /**< Reflected inertia matrix  */
+    bool m_useReflectedInertia; /**< If true the reflected inertia will be used  */
+
+public:
+    /**
+     * Constructor. If you call this constructor the motor reflected inertia is disabled
+     * @param kinDyn an iDynTree kinDyn computation object
+     * @param handler the variable handler object
+     * @param framesInContact vector containing the frames in contact. Each element of the vector is
+     * a pair containing the name of the frame in the variableHandler and in the model.
+     * @throw std::runtime_error if the frame is not defined
+     */
+    SystemDynamicsElement(iDynTree::KinDynComputations& kinDyn,
+                          const VariableHandler& handler,
+                          const std::vector<std::pair<std::string, std::string>>& framesInContact);
```

Comment:
This is not clear to me. Why a pair of strings? It is super easy to confuse what is the first or the second.

Reply:
The first element of the pair is the name used to indicate the name of the variable associated to the frame in the `handler` while the second element is the name of the frame in the model. 
For instance, the first element of the pair could be `left foot` and the second one `l_sole`. 
However, I agree with you. As @traversaro suggested I should avoid using `std::pair<>`.
> I personally don't like the use of pair for this kind of uses, because then you access the elements as bounds.first and bounds.second, and that is an harm to readability.
> IMHO a custom template pair-structure with min and max is more readable , see https://arne-mertz.de/2017/03/smelly-pair-tuple/ .

Probably it is easier to discuss F2F